embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/telemetry/EmbraceTelemetryService.kt

```diff
@@ -0,0 +1,77 @@
+package io.embrace.android.embracesdk.telemetry
+
+import io.embrace.android.embracesdk.capture.metadata.EmbraceMetadataService
+import io.embrace.android.embracesdk.internal.spans.toEmbraceAttributeName
+import java.util.Locale
+
+/*
+    Service for tracking usage of public APIs, and different internal metrics about the app.
+ */
+internal class EmbraceTelemetryService(
+    private val okHttpReflectionFacade: OkHttpReflectionFacade = OkHttpReflectionFacade()
+) {
+
+    private val usageCountMap = mutableMapOf<String, Int>()
+    private val appAttributesMap = mutableMapOf<String, String>()
+
+    /*
+        Tracks the usage of a public API by name. Adds a suffix for easier identification. Replaces whitespaces with underscores.
+    */
+    fun onPublicApiCalled(name: String) {
+        val suffixedName = "emb.usage.${
+        name.toLowerCase(Locale.ENGLISH)
+            .trim()
+            .replace(" ", "_")
+        }"
```

Comment:
Because this function is on a hot execution path it would be ideal to avoid all these operations (changing case, creating string subsequences, etc).

Could we either alter the code so that `onPublicApiCalled` always receives a sanitized string, or defer the sanitisation to the point where spans are flushed?

Reply:
I thought it was going to be better to do these operations gradually rather than all at the same time 😅

Either way, I'll modify calls to `checkSdkStartedAndLogPublicApiUsage("set user identifier")` to sanitize them.

It will only affect `logSDKNotInitialized()`, so it wouldn't be too inconvenient.

Reply:
It's an interesting point. Something like `logBreadcrumb` might get spammed many times a second, but session serialization also happens every ~2s. I think it'd be preferable to avoid sanitisation completely & just pass in something that we can use as-is. e.g. `setUserIdentifier` is good enough for both uses

Reply:
We don't do this on session serialization! `flushSpans` only happens when the session actually ends, so the frequency is lower. 

I think your way is better though: if we sanitize when we add it to the map, we may sanitize the same string multiple times (multiple calls to logBreadcrumb for instance). If we do it at the end, we are going to do it just once for every String! 

I still have to append the `emb.usage.` prefix, so I'll do it on getUsageTelemetry

Reply:
Another option you can consider is creating a static map of the method name to the sanitized name and have it defined there. Having a map like that is nice because we can use it in a test to verify that each method has been called in `NullParametersTests` and others like that, which we have no good way to verify right now. We can also make it a failing condition in another test that all public methods go through this helper so new methods are handled appropriately.

If you go with this suggestion, all that other stuff I mention will of course be another PR - probably a KTLO task we'll prioritize separately. Totally OK not to go with that as well as what y'all are talking about seem fine too